### PR TITLE
MaybeUninit: some docs, rename into_inner -> into_initialized, return &mut from set

### PR DESCRIPTION
**Pull request by RalfJung**

### CODE REVIEW

File: src/libcore/mem.rs [hunk 1]
```diff
@@ -1034,7 +1034,41 @@ impl<T: ?Sized> DerefMut for ManuallyDrop<T> {
...
+/// be aligned and non-NULL.  This is an invariant that must *always* be upheld,
+/// even in unsafe code.  As a consequence, 0-initializing a variable of reference
+/// type causes instantaneous undefined behavior, no matter whether that reference
+/// ever gets used to access memory:
+/// ```rust,ignore
+/// use std::mem;
+///
+/// let x: &i32 = mem::zeroed(); // undefined behavior!
+/// ```
+/// This is exploitet by the compiler for various optimizations, such as eliding
```
Comment (memoryruins) [inline src/libcore/mem.rs]:
```suggestion
/// This is exploited by the compiler for various optimizations, such as eliding
```

File: src/libcore/mem.rs [hunk 2]
```diff
@@ -1034,7 +1034,41 @@ impl<T: ?Sized> DerefMut for ManuallyDrop<T> {
...
+/// run-time checks and optimizing `enum` layout.
+///
+/// Not initializing memory at all (instead of 0-initializing it) causes the same
+/// issue: after all, the initial value of the variable might just happen to be
+/// one that violates the invariant.
+///
+/// `MaybeUninit` serves to enable unsafe code to deal with uninitialized data:
+/// it is a signal to the compiler indicating that the data here may *not*
```
Comment (memoryruins) [inline src/libcore/mem.rs]:
```suggestion
/// it is a signal to the compiler indicating that the data here might *not*
```

Comment (RalfJung) [reply src/libcore/mem.rs]: Thanks, fixed!

File: src/libcore/mem.rs (hunk 1 above)
Comment (RalfJung) [reply src/libcore/mem.rs]: Thanks, fixed!

**Comment by cramertj** [comment]
r=me w/ travis

**Comment by RalfJung** [comment]
This fails on many but not all platforms
```
00:24:26] error: method is never used: `into_inner`
[00:24:26]     --> src/libcore/mem.rs:1151:5
[00:24:26]      |
[00:24:26] 1151 |     pub(crate) unsafe fn into_inner(self) -> T {
[00:24:26]      |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[00:24:26]      |
[00:24:26]      = note: `-D dead-code` implied by `-D warnings`
```

@bors r-

**Comment by RalfJung** [comment]
I think this should fix the SGX failures, but I don't know how to test that locally...

@bors r=cramertj

**Pull request merged**